edgedb-client/src/builder.rs

```diff
@@ -481,13 +517,55 @@ impl Builder {
         self.password = credentials.password.clone();
         self.database = credentials.database.clone()
                 .unwrap_or_else(|| "edgedb".into());
-        self.verify_hostname = credentials.tls_verify_hostname;
+        self.creds_file_outdated = credentials.tls_security.is_none() &&
+            credentials.tls_verify_hostname.is_some();
+        let mut expected_verify_hostname = None;
+        self.client_security = match credentials.tls_security.as_deref() {
+            None => match credentials.tls_verify_hostname {
+                None => TlsClientSecurity::Default,
+                Some(true) => TlsClientSecurity::Strict,
+                Some(false) => TlsClientSecurity::NoHostVerification,
+            },
+            Some("default") => TlsClientSecurity::Default,
+            Some("strict") => {
+                expected_verify_hostname = Some(true);
+                TlsClientSecurity::Strict
+            },
+            Some("no_verify_hostname") => {
+                expected_verify_hostname = Some(false);
+                TlsClientSecurity::NoHostVerification
+            },
+            Some("insecure") => TlsClientSecurity::Insecure,
+            Some(tls_security) => {
+                return Err(ClientError::with_message(format!(
+                    "Invalid value `{}` for `tls_security`. \
+                     Options: default, insecure, \
+                     no_host_verification, strict.",
+                    tls_security,
+                )));
+            }
+        };
+        if let Some(tls_verify_hostname) = credentials.tls_verify_hostname {
+            if let Some(tls_security) = &credentials.tls_security {
+                if expected_verify_hostname != Some(tls_verify_hostname) {
+                    return Err(ClientError::with_message(format!(
+                        "tls_security={} but tls_verify_hostname={:?}",
+                        tls_security, tls_verify_hostname,
+                    )));
+                }
+            }
+        }
         self.cert = cert;
         self.pem = pem;
         self.initialized = true;
         Ok(self)
     }
 
+    /// Returns if the credentials file should be updated.
+    pub fn should_update_credential_file(&self) -> bool {
```

Comment:
This should be hidden under `#[cfg(feature="unstable")]` so it's only available to the CLI.